{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alternative Splicing in MD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from transvis.io import *\n",
    "from transvis.alterSplice import *\n",
    "from transvis.utils import *\n",
    "initDB()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# you can set this up to 2**64 if you want everything processed, or less for debugging.\n",
    "processedNo = 2**64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from intervaltree import Interval, IntervalTree\n",
    "intervaltrees = {}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We're going to iterate through all probesets, and retrive from the annotation metadata corresponding to each probeset the probeset location, expressed in genomic coordinates with GRCh37 as the reference genome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Takes about 5 minutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "currentIter = transIter()\n",
    "\n",
    "allchr = set()\n",
    "\n",
    "# ml := metadataLookup\n",
    "ml = {key : value for key, value in zip(metadataKeys(), range(len(metadataKeys())))}\n",
    "for _, transID in zip(range(processedNo), currentIter):\n",
    "    probesets = transToProbeset(transID)\n",
    "    for probeset in probesets:\n",
    "        metadata = probesetAnnotationMetadata(probeset)\n",
    "        seqname = metadata[ml['seqname']]\n",
    "        strand = metadata[ml['strand']]\n",
    "        seqname += strand\n",
    "        if re.findall(\"^chr*\", seqname):\n",
    "            allchr.add(seqname)\n",
    "            if seqname not in intervaltrees:\n",
    "                t = IntervalTree()\n",
    "                intervaltrees[seqname] = t\n",
    "            else:\n",
    "                t = intervaltrees[seqname]\n",
    "            start = metadata[ml['start']]\n",
    "            stop = metadata[ml['stop']]\n",
    "            try:\n",
    "                start = int(start)\n",
    "                stop = int(stop)\n",
    "            except ValueError:\n",
    "                continue\n",
    "            if start == 0 or stop == 0:\n",
    "                continue\n",
    "            t[start:stop] = probeset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Then, we'll iterate through all genecode transcripts, and use the Binary Interval Tree, to find those probesets, which are strictly contained within exons belonging to these transcripts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Takes another 5 minutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for c in allchr:\n",
    "    c = c[:-1]\n",
    "    for genecodeI, genecode in zip(range(processedNo), genecodeIter(c)):\n",
    "        for transcriptI, transcript in enumerate(genecode):\n",
    "            typ = transcript[0]\n",
    "            chromosome = transcript[1]\n",
    "            suffix = transcript[2]\n",
    "            seqname = chromosome + suffix\n",
    "            if typ == \"exon\":\n",
    "                key = transcript[-1]\n",
    "                start = int(transcript[3])\n",
    "                stop = int(transcript[4])\n",
    "                setGenecodeGenenameBoundary(key, chromosome, suffix, start, stop)\n",
    "                result = intervaltrees[seqname].search(start, stop, strict=True)\n",
    "                if result:\n",
    "                    for interval in result:\n",
    "                        probeset = interval.data\n",
    "                        setGenecodeGenenameToProbeset(key, probeset)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For example all MBNL1's exons strictly contain the following probesets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'2648142',\n",
       " '2648143',\n",
       " '2648146',\n",
       " '2648147',\n",
       " '2648158',\n",
       " '2648159',\n",
       " '2648160',\n",
       " '2648161',\n",
       " '2648167',\n",
       " '2648174',\n",
       " '2648175',\n",
       " '2648178',\n",
       " '2648180',\n",
       " '2648181',\n",
       " '2648184',\n",
       " '2648185',\n",
       " '2648186',\n",
       " '2648194',\n",
       " '2648197',\n",
       " '2648199',\n",
       " '2648200',\n",
       " '2648201',\n",
       " '2648202',\n",
       " '2648203',\n",
       " '2648204',\n",
       " '2648205'}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from transvis.io import *\n",
    "from transvis.alterSplice import *\n",
    "from transvis.utils import *\n",
    "initDB()\n",
    "getGenecodeGenenameToProbeset(\"MBNL1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For the visualisation tool we also record the genomic coordinates of each gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[151961617, 152183569, 'chr3', '+']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "getGenecodeGenenameBoundary(\"MBNL1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "genecodeHGNC = set(d(i).split(\"$\")[-1] for i in redisConn[\"genecode\"].keys(\"genename$probeset$*\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "properHGNC = set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open(\"CEL_files/HGNC_names.txt\") as f:\n",
    "    for line in f:\n",
    "        properHGNC.add(line.strip())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The file HGNC_names.txt contains all legal HGNC names, as per the HGNC website. As one can see, there are 40938 genes in total in this reference database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40938"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(properHGNC)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Our method of using genecode identifies over 43 thousand genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "43705"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genecodeHGNC)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Let us narrow down to the genes at the intersection of these gene lists. (potentialy caveat: genecode could be using a slightly different nomenclature...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "genesToInvest = properHGNC.intersection(genecodeHGNC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28898"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genesToInvest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "distribution = [len(getGenecodeGenenameToProbeset(i)) for i in genesToInvest]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len([i for i in distribution if i > 400])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTN\n"
     ]
    }
   ],
   "source": [
    "for i in genesToInvest:\n",
    "    probesets = getGenecodeGenenameToProbeset(i)\n",
    "    if len(probesets) > 400:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# That for example is a histogram of probesets/gene. The gene with the highest number of probesets is titin (which is actually a very interesting gene/protein, and a sensible answer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD8CAYAAAB+UHOxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFKVJREFUeJzt3X+snuV93/H3pziQH52wgTOL2c5MFasRmRbCjsBRqqmD\nxRioYv6giKgaFrLk/cG2ZKrUmU2aVUgkkKbSIC1oVvFqoiyE0kRYgEo9QzXtD34cAiX8KPMJgWIL\n8Ak2zlpUVtPv/nguu0+cc3KeY58f4Ov9kh491/29r/t+rvsOOR/fP57nTlUhSerPLy31ACRJS8MA\nkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHVq2VIP4Bc577zzau3atUs9DEn6UHn6\n6ad/UlVjs/X7QAfA2rVrmZiYWOphSNKHSpLXRunnKSBJ6tRIAZDk3yd5IcnzSb6T5KNJLkjyRJLJ\nJN9Ncmbre1abnmzz1w6t5+ZWfznJFQuzSZKkUcwaAElWAf8OGK+qfwKcAVwP3A7cUVWfAg4DW9oi\nW4DDrX5H60eSC9tynwE2At9Mcsb8bo4kaVSjngJaBnwsyTLg48AbwGXA/W3+LuCa1t7UpmnzL0+S\nVr+3qt6rqh8Dk8Alp74JkqSTMWsAVNUB4L8Af8ngD/8R4Gngnao62rrtB1a19irg9bbs0db/3OH6\nNMscl2RrkokkE1NTUyezTZKkEYxyCmgFg3+9XwD8I+ATDE7hLIiq2lFV41U1PjY2611MkqSTNMop\noH8J/Liqpqrqb4HvAV8AlrdTQgCrgQOtfQBYA9Dmnw28PVyfZhlJ0iIbJQD+Elif5OPtXP7lwIvA\nY8C1rc9m4IHW3t2mafMfrcFzJ3cD17e7hC4A1gFPzs9mSJLmatYvglXVE0nuB34AHAWeAXYADwH3\nJvlaq93dFrkb+FaSSeAQgzt/qKoXktzHIDyOAjdV1fvzvD2SpBHlg/xQ+PHx8TqVbwKv3fbQ8far\nt109H0OSpA+8JE9X1fhs/fwmsCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAk\ndcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHVq1gBI8qtJnh16/TTJV5Ock2RP\nkn3tfUXrnyR3JplM8lySi4fWtbn135dk88yfKklaaLMGQFW9XFUXVdVFwD8D3gW+D2wD9lbVOmBv\nmwa4ksED39cBW4G7AJKcA2wHLgUuAbYfCw1J0uKb6ymgy4EfVdVrwCZgV6vvAq5p7U3APTXwOLA8\nyfnAFcCeqjpUVYeBPcDGU94CSdJJmWsAXA98p7VXVtUbrf0msLK1VwGvDy2zv9VmqkuSlsDIAZDk\nTOBLwB+dOK+qCqj5GFCSrUkmkkxMTU3NxyolSdOYyxHAlcAPquqtNv1WO7VDez/Y6geANUPLrW61\nmeo/o6p2VNV4VY2PjY3NYXiSpLmYSwB8mb8//QOwGzh2J89m4IGh+g3tbqD1wJF2qugRYEOSFe3i\n74ZWkyQtgWWjdEryCeCLwL8eKt8G3JdkC/AacF2rPwxcBUwyuGPoRoCqOpTkVuCp1u+Wqjp0ylsg\nSTopIwVAVf01cO4JtbcZ3BV0Yt8CbpphPTuBnXMfpiRpvvlNYEnqlAEgSZ0yACSpUwaAJHXKAJCk\nThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqU\nASBJnRopAJIsT3J/kr9I8lKSzyc5J8meJPva+4rWN0nuTDKZ5LkkFw+tZ3Prvy/J5pk/UZK00EY9\nAvgG8CdV9Wngs8BLwDZgb1WtA/a2aYArgXXttRW4CyDJOcB24FLgEmD7sdCQJC2+WQMgydnAPwfu\nBqiq/1dV7wCbgF2t2y7gmtbeBNxTA48Dy5OcD1wB7KmqQ1V1GNgDbJzXrZEkjWyUI4ALgCngvyd5\nJskfJPkEsLKq3mh93gRWtvYq4PWh5fe32kz1n5Fka5KJJBNTU1Nz2xpJ0shGCYBlwMXAXVX1OeCv\n+fvTPQBUVQE1HwOqqh1VNV5V42NjY/OxSknSNEYJgP3A/qp6ok3fzyAQ3mqndmjvB9v8A8CaoeVX\nt9pMdUnSEpg1AKrqTeD1JL/aSpcDLwK7gWN38mwGHmjt3cAN7W6g9cCRdqroEWBDkhXt4u+GVpMk\nLYFlI/b7t8C3k5wJvALcyCA87kuyBXgNuK71fRi4CpgE3m19qapDSW4Fnmr9bqmqQ/OyFZKkORsp\nAKrqWWB8mlmXT9O3gJtmWM9OYOdcBihJWhh+E1iSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQ\npE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdWrU5wF86K3d9tDx9qu3Xb2EI5GkDwaP\nACSpUwaAJHVqpABI8mqSHyZ5NslEq52TZE+Sfe19RasnyZ1JJpM8l+TiofVsbv33Jdk80+dJkhbe\nXI4A/kVVXVRVxx4NuQ3YW1XrgL1tGuBKYF17bQXugkFgANuBS4FLgO3HQkOStPhO5RTQJmBXa+8C\nrhmq31MDjwPLk5wPXAHsqapDVXUY2ANsPIXPlySdglEDoIA/TfJ0kq2ttrKq3mjtN4GVrb0KeH1o\n2f2tNlNdkrQERr0N9Neq6kCSfwjsSfIXwzOrqpLUfAyoBcxWgE9+8pPzsUpJ0jRGOgKoqgPt/SDw\nfQbn8N9qp3Zo7wdb9wPAmqHFV7faTPUTP2tHVY1X1fjY2NjctkaSNLJZAyDJJ5L8g2NtYAPwPLAb\nOHYnz2bggdbeDdzQ7gZaDxxpp4oeATYkWdEu/m5oNUnSEhjlFNBK4PtJjvX/H1X1J0meAu5LsgV4\nDbiu9X8YuAqYBN4FbgSoqkNJbgWeav1uqapD87YlkqQ5mTUAquoV4LPT1N8GLp+mXsBNM6xrJ7Bz\n7sOUJM03vwksSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1\nygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnRo5AJKckeSZJA+26QuSPJFkMsl3k5zZ6me16ck2\nf+3QOm5u9ZeTXDHfGyNJGt1cjgC+Arw0NH07cEdVfQo4DGxp9S3A4Va/o/UjyYXA9cBngI3AN5Oc\ncWrDlySdrJECIMlq4GrgD9p0gMuA+1uXXcA1rb2pTdPmX976bwLurar3qurHDB4af8l8bIQkae5G\nPQL4feB3gL9r0+cC71TV0Ta9H1jV2quA1wHa/COt//H6NMtIkhbZrAGQ5DeAg1X19CKMhyRbk0wk\nmZiamlqMj5SkLo1yBPAF4EtJXgXuZXDq5xvA8iTLWp/VwIHWPgCsAWjzzwbeHq5Ps8xxVbWjqsar\nanxsbGzOGyRJGs2sAVBVN1fV6qpay+Ai7qNV9VvAY8C1rdtm4IHW3t2mafMfrapq9evbXUIXAOuA\nJ+dtSyRJc7Js9i4z+g/AvUm+BjwD3N3qdwPfSjIJHGIQGlTVC0nuA14EjgI3VdX7p/D5kqRTMKcA\nqKo/A/6stV9hmrt4qupvgN+cYfmvA1+f6yAlSfPPbwJLUqcMAEnqlAEgSZ0yACSpUwaAJHXqVG4D\nPS2s3fbQ8fart129hCORpMXlEYAkdcoAkKROGQCS1CkDQJI61eVF4OELv5LUK48AJKlTBoAkdcoA\nkKROGQCS1CkDQJI6ZQBIUqcMAEnq1KwBkOSjSZ5M8udJXkjyu61+QZInkkwm+W6SM1v9rDY92eav\nHVrXza3+cpIrFmqjJEmzG+UI4D3gsqr6LHARsDHJeuB24I6q+hRwGNjS+m8BDrf6Ha0fSS5k8ID4\nzwAbgW8mOWM+N0aSNLpZA6AG/qpNfqS9CrgMuL/VdwHXtPamNk2bf3mStPq9VfVeVf0YmGSah8pL\nkhbHSNcAkpyR5FngILAH+BHwTlUdbV32A6taexXwOkCbfwQ4d7g+zTKSpEU2UgBU1ftVdRGwmsG/\n2j+9UANKsjXJRJKJqamphfoYSerenO4Cqqp3gMeAzwPLkxz7MbnVwIHWPgCsAWjzzwbeHq5Ps8zw\nZ+yoqvGqGh8bG5vL8CRJczDKXUBjSZa39seALwIvMQiCa1u3zcADrb27TdPmP1pV1erXt7uELgDW\nAU/O14ZIkuZmlJ+DPh/Y1e7Y+SXgvqp6MMmLwL1JvgY8A9zd+t8NfCvJJHCIwZ0/VNULSe4DXgSO\nAjdV1fvzuzmSpFHNGgBV9RzwuWnqrzDNXTxV9TfAb86wrq8DX5/7MCVJ863LB8LMZPhBMa/edvUS\njkSSFp4/BSFJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXK\nAJCkThkAktQpA0CSOmUASFKnDABJ6tQozwRek+SxJC8meSHJV1r9nCR7kuxr7ytaPUnuTDKZ5Lkk\nFw+ta3Prvy/J5pk+U5K08EY5AjgK/HZVXQisB25KciGwDdhbVeuAvW0a4EoGD3xfB2wF7oJBYADb\ngUsZPEpy+7HQkCQtvlkDoKreqKoftPb/BV4CVgGbgF2t2y7gmtbeBNxTA48Dy5OcD1wB7KmqQ1V1\nGNgDbJzXrZEkjWxO1wCSrGXwgPgngJVV9Uab9SawsrVXAa8PLba/1WaqS5KWwMgBkOSXgT8GvlpV\nPx2eV1UF1HwMKMnWJBNJJqampuZjlZKkaYwUAEk+wuCP/7er6nut/FY7tUN7P9jqB4A1Q4uvbrWZ\n6j+jqnZU1XhVjY+Njc1lWyRJczDKXUAB7gZeqqrfG5q1Gzh2J89m4IGh+g3tbqD1wJF2qugRYEOS\nFe3i74ZWkyQtgWUj9PkC8K+AHyZ5ttX+I3AbcF+SLcBrwHVt3sPAVcAk8C5wI0BVHUpyK/BU63dL\nVR2al62QJM3ZrAFQVf8byAyzL5+mfwE3zbCuncDOuQxQkrQw/CawJHXKAJCkThkAktQpA0CSOmUA\nSFKnRrkNtEtrtz10vP3qbVcv4UgkaWF4BCBJnTIAJKlTBoAkdcoAkKROGQCS1CnvAhqBdwRJOh15\nBCBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6NcpD4XcmOZjk+aHaOUn2JNnX3le0epLcmWQyyXNJ\nLh5aZnPrvy/J5uk+S5K0eEY5AvhDYOMJtW3A3qpaB+xt0wBXAuvaaytwFwwCA9gOXApcAmw/FhqS\npKUxawBU1f8CDp1Q3gTsau1dwDVD9Xtq4HFgeZLzgSuAPVV1qKoOA3v4+VCRJC2ik70GsLKq3mjt\nN4GVrb0KeH2o3/5Wm6n+c5JsTTKRZGJqauokhydJms0pXwSuqgJqHsZybH07qmq8qsbHxsbma7WS\npBOcbAC81U7t0N4PtvoBYM1Qv9WtNlNdkrRETjYAdgPH7uTZDDwwVL+h3Q20HjjSThU9AmxIsqJd\n/N3QapKkJTLrr4Em+Q7w68B5SfYzuJvnNuC+JFuA14DrWveHgauASeBd4EaAqjqU5Fbgqdbvlqo6\n8cKyJGkRZXAK/4NpfHy8JiYmTnr54Z9xXgj+NLSkD6IkT1fV+Gz9/CawJHXKAJCkThkAktQpHwl5\nCnxUpKQPM48AJKlTBoAkdcoAkKROGQCS1CkDQJI65V1A88Q7giR92HgEIEmdMgAkqVMGgCR1ygCQ\npE55EXgBeEFY0oeBRwCS1CmPABaYRwOSPqgW/QggycYkLyeZTLJtsT9fkjSwqEcASc4A/ivwRWA/\n8FSS3VX14mKOY6nM9IhKjwwkLYXFPgV0CTBZVa8AJLkX2AR0EQAz8TSRpKWw2AGwCnh9aHo/cOki\nj+EDbb4eZH9ikBgykk70gbsInGQrsLVN/lWSl09yVecBP5mfUX345PbjzZ/bD0PzetL1fw9D3A8D\np/t++MejdFrsADgArBmaXt1qx1XVDmDHqX5QkomqGj/V9XzYuR8G3A8D7ocB98PAYt8F9BSwLskF\nSc4Ergd2L/IYJEks8hFAVR1N8m+AR4AzgJ1V9cJijkGSNLDo1wCq6mHg4UX4qFM+jXSacD8MuB8G\n3A8D7gcgVbXUY5AkLQF/C0iSOnXaBUBvPzWRZGeSg0meH6qdk2RPkn3tfUWrJ8mdbd88l+TipRv5\n/EmyJsljSV5M8kKSr7R6b/vho0meTPLnbT/8bqtfkOSJtr3fbTdgkOSsNj3Z5q9dyvHPtyRnJHkm\nyYNtusv98IucVgEw9FMTVwIXAl9OcuHSjmrB/SGw8YTaNmBvVa0D9rZpGOyXde21Fbhrkca40I4C\nv11VFwLrgZva/+697Yf3gMuq6rPARcDGJOuB24E7qupTwGFgS+u/BTjc6ne0fqeTrwAvDU33uh9m\nVlWnzQv4PPDI0PTNwM1LPa5F2O61wPND0y8D57f2+cDLrf3fgC9P1+90egEPMPi9qW73A/Bx4AcM\nvmn/E2BZqx///wiDu/E+39rLWr8s9djnaftXMwj9y4AHgfS4H2Z7nVZHAEz/UxOrlmgsS2llVb3R\n2m8CK1v7tN8/7fD9c8ATdLgf2mmPZ4GDwB7gR8A7VXW0dRne1uP7oc0/Apy7uCNeML8P/A7wd236\nXPrcD7/Q6RYAOkEN/lnTxa1eSX4Z+GPgq1X10+F5veyHqnq/qi5i8C/gS4BPL/GQFl2S3wAOVtXT\nSz2WD7rTLQBm/amJTryV5HyA9n6w1U/b/ZPkIwz++H+7qr7Xyt3th2Oq6h3gMQanOpYnOfadn+Ft\nPb4f2vyzgbcXeagL4QvAl5K8CtzL4DTQN+hvP8zqdAsAf2piYDewubU3Mzgnfqx+Q7sLZj1wZOgU\nyYdWkgB3Ay9V1e8NzeptP4wlWd7aH2NwHeQlBkFwbet24n44tn+uBR5tR0ofalV1c1Wtrqq1DP4G\nPFpVv0Vn+2EkS30RYr5fwFXA/2Fw7vM/LfV4FmF7vwO8Afwtg/OaWxicv9wL7AP+J3BO6xsGd0n9\nCPghML7U45+nffBrDE7vPAc8215Xdbgf/inwTNsPzwP/udV/BXgSmAT+CDir1T/apifb/F9Z6m1Y\ngH3y68CDve+HmV5+E1iSOnW6nQKSJI3IAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVP/\nHzEm2jAMTJwhAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fa972795400>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ax1 = plt.subplot()\n",
    "_ = ax1.hist(distribution, bins=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# We can zoom in to discover how many genes have outragesouly few probes (I suggest we resign from all genes with < 4 probes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAD8CAYAAACVZ8iyAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAENpJREFUeJzt3X+s3XV9x/Hna61zijpxvSOlLbloCktpZtEbZHMaNpwU\ndBaXhbXbEBmzEmHDzcSAW4JxYyGb6DRzXap0QIZFBjqaWX9UZiRLBnoLDRSQeYEi9+7SXkXB6IIC\n7/1xvqWH9rb3xznc74U+H8nJ+Z73+f54n28or/v9fj/fc1JVSJIObz/XdgOSpPYZBpIkw0CSZBhI\nkjAMJEkYBpIkDANJEoaBJAnDQJIELGy7gaksWrSoBgcH225Dkp43tm/f/r2qGpjJMvM+DAYHBxke\nHm67DUl63kjy0EyX8TSRJMkwkCQZBpIkDANJEtMIgySbkuxJsrOr9rkkO5rHriQ7mvpgkv/reu+f\nu5Z5fZK7kowk+WSSPDcfSZI0U9MZTXQV8I/ANXsLVfX7e6eTXAE81jX//VW1apL1bADeA9wGbAVW\nA1+aecuSpH6b8sigqm4BHp3sveav+7OAzYdaR5LFwCuq6tbq/LTaNcCZM29XkvRc6PWawZuA3VX1\nna7asUnuSPKNJG9qakuA0a55RpvapJKsTzKcZHhiYqLHFiVJU+k1DNbx7KOCceCYqjoR+Avgs0le\nMdOVVtXGqhqqqqGBgRndRCdJmoVZ34GcZCHwu8Dr99aq6gngiWZ6e5L7geOAMWBp1+JLm5qmMHjx\nF2e97K7L39bHTiS9kPVyZPAW4NtV9czpnyQDSRY0068GlgMPVNU48HiSk5vrDO8Cbuph25KkPprO\n0NLNwH8DxycZTXJe89ZaDrxw/Gbgzmao6Q3A+VW19+Lz+4DPACPA/TiSSJLmjSlPE1XVuoPU3z1J\n7UbgxoPMPwysnGF/kqQ54B3IkiTDQJL0PPg9A7XHkUzS4cMjA0mSYSBJMgwkSRgGkiQMA0kSjiaa\nFkfVSHqh88hAkmQYSJIMA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEtMIgySbkuxJ\nsrOr9uEkY0l2NI8zut67JMlIkvuSnNZVX93URpJc3P+PIkmarekcGVwFrJ6k/vGqWtU8tgIkWQGs\nBU5olvmnJAuSLAA+BZwOrADWNfNKkuaBKb+1tKpuSTI4zfWtAa6rqieAB5OMACc1741U1QMASa5r\n5r1nxh1Lkvqul2sGFya5szmNdGRTWwI83DXPaFM7WF2SNA/MNgw2AK8BVgHjwBV96whIsj7JcJLh\niYmJfq5akjSJWYVBVe2uqqeq6mng0+w7FTQGLOuadWlTO1j9YOvfWFVDVTU0MDAwmxYlSTMwq186\nS7K4qsabl+8E9o402gJ8NsnHgKOB5cA3gQDLkxxLJwTWAn/QS+N6YfPX5aS5NWUYJNkMnAIsSjIK\nXAqckmQVUMAu4L0AVXV3kuvpXBh+Erigqp5q1nMh8BVgAbCpqu7u+6eRJM3KdEYTrZukfOUh5r8M\nuGyS+lZg64y6kyTNCe9AliQZBpIkw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkoRh\nIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJAnDQJKEYSBJwjCQJAELp5ohySbg7cCeqlrZ1P4e+B3g\np8D9wLlV9cMkg8C9wH3N4rdW1fnNMq8HrgJeAmwFLqqq6ueHkfph8OIvznrZXZe/rY+dSHNnOkcG\nVwGr96ttA1ZW1a8C/wNc0vXe/VW1qnmc31XfALwHWN489l+nJKklU4ZBVd0CPLpf7atV9WTz8lZg\n6aHWkWQx8IqqurU5GrgGOHN2LUuS+q0f1wz+GPhS1+tjk9yR5BtJ3tTUlgCjXfOMNjVJ0jww5TWD\nQ0nyl8CTwLVNaRw4pqq+31wj+PckJ8xiveuB9QDHHHNMLy1KkqZh1kcGSd5N58LyH+69EFxVT1TV\n95vp7XQuLh8HjPHsU0lLm9qkqmpjVQ1V1dDAwMBsW5QkTdOswiDJauCDwDuq6idd9YEkC5rpV9O5\nUPxAVY0Djyc5OUmAdwE39dy9JKkvpjO0dDNwCrAoyShwKZ3RQy8GtnX+3/7MENI3Ax9J8jPgaeD8\nqtp78fl97Bta+iWefZ1BktSiKcOgqtZNUr7yIPPeCNx4kPeGgZUz6k6SNCe8A1mS1NtoIkn908ud\nz+Ddz+qNRwaSJMNAkmQYSJIwDCRJGAaSJAwDSRKGgSQJ7zOQ1Af+Otzzn0cGkiTDQJJkGEiSMAwk\nSRgGkiQMA0kShoEkCcNAkoRhIEnCMJAkYRhIkpjmdxMl2QS8HdhTVSub2quAzwGDwC7grKr6QZIA\nnwDOAH4CvLuqbm+WOQf4q2a1f1NVV/fvo0g6HPm9SP0x3SODq4DV+9UuBm6uquXAzc1rgNOB5c1j\nPbABngmPS4E3ACcBlyY5spfmJUn9Ma0wqKpbgEf3K68B9v5lfzVwZlf9muq4FXhlksXAacC2qnq0\nqn4AbOPAgJEktaCXawZHVdV4M/0IcFQzvQR4uGu+0aZ2sLokqWV9uYBcVQVUP9YFkGR9kuEkwxMT\nE/1arSTpIHoJg93N6R+a5z1NfQxY1jXf0qZ2sPoBqmpjVQ1V1dDAwEAPLUqSpqOXXzrbApwDXN48\n39RVvzDJdXQuFj9WVeNJvgL8bddF47cCl/SwfUlqTS+jmGD+jWSa7tDSzcApwKIko3RGBV0OXJ/k\nPOAh4Kxm9q10hpWO0Blaei5AVT2a5K+BbzXzfaSq9r8oLUlqwbTCoKrWHeStUyeZt4ALDrKeTcCm\naXcnSZoT3oEsSTIMJEmGgSSJ3kYTSZJa0OtIpsl4ZCBJMgwkSYaBJAnDQJKEYSBJwjCQJGEYSJIw\nDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiR6CIMkxyfZ0fV4PMn7k3w4yVhX\n/YyuZS5JMpLkviSn9ecjSJJ6Neufvayq+4BVAEkWAGPAF4BzgY9X1Ue750+yAlgLnAAcDXwtyXFV\n9dRse5Ak9Ue/ThOdCtxfVQ8dYp41wHVV9URVPQiMACf1afuSpB70KwzWApu7Xl+Y5M4km5Ic2dSW\nAA93zTPa1A6QZH2S4STDExMTfWpRknQwPYdBkp8H3gH8W1PaALyGzimkceCKma6zqjZW1VBVDQ0M\nDPTaoiRpCv04MjgduL2qdgNU1e6qeqqqngY+zb5TQWPAsq7lljY1SVLL+hEG6+g6RZRkcdd77wR2\nNtNbgLVJXpzkWGA58M0+bF+S1KNZjyYCSHIE8NvAe7vKf5dkFVDArr3vVdXdSa4H7gGeBC5wJJEk\nzQ89hUFV/Rj4pf1qZx9i/suAy3rZpiSp/7wDWZJkGEiSDANJEoaBJAnDQJKEYSBJwjCQJGEYSJIw\nDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQMA0kShoEkCcNAkkQfwiDJriR3\nJdmRZLipvSrJtiTfaZ6PbOpJ8skkI0nuTPK6XrcvSepdv44MfrOqVlXVUPP6YuDmqloO3Ny8Bjgd\nWN481gMb+rR9SVIPnqvTRGuAq5vpq4Ezu+rXVMetwCuTLH6OepAkTVM/wqCArybZnmR9Uzuqqsab\n6UeAo5rpJcDDXcuONrVnSbI+yXCS4YmJiT60KEk6lIV9WMdvVNVYkl8GtiX5dvebVVVJaiYrrKqN\nwEaAoaGhGS0rSZq5no8Mqmqsed4DfAE4Cdi99/RP87ynmX0MWNa1+NKmJklqUU9hkOSIJC/fOw28\nFdgJbAHOaWY7B7ipmd4CvKsZVXQy8FjX6SRJUkt6PU10FPCFJHvX9dmq+nKSbwHXJzkPeAg4q5l/\nK3AGMAL8BDi3x+1LkvqgpzCoqgeA105S/z5w6iT1Ai7oZZuSpP7zDmRJkmEgSTIMJEkYBpIkDANJ\nEoaBJAnDQJKEYSBJwjCQJGEYSJIwDCRJGAaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgG\nkiR6CIMky5J8Pck9Se5OclFT/3CSsSQ7mscZXctckmQkyX1JTuvHB5Ak9W5hD8s+CXygqm5P8nJg\ne5JtzXsfr6qPds+cZAWwFjgBOBr4WpLjquqpHnqQJPXBrI8Mqmq8qm5vpn8E3AssOcQia4DrquqJ\nqnoQGAFOmu32JUn905drBkkGgROB25rShUnuTLIpyZFNbQnwcNdioxw6PCRJc6TnMEjyMuBG4P1V\n9TiwAXgNsAoYB66YxTrXJxlOMjwxMdFri5KkKfQUBkleRCcIrq2qzwNU1e6qeqqqngY+zb5TQWPA\nsq7Flza1A1TVxqoaqqqhgYGBXlqUJE1DL6OJAlwJ3FtVH+uqL+6a7Z3AzmZ6C7A2yYuTHAssB745\n2+1Lkvqnl9FEbwTOBu5KsqOpfQhYl2QVUMAu4L0AVXV3kuuBe+iMRLrAkUSSND/MOgyq6r+ATPLW\n1kMscxlw2Wy3KUl6bngHsiTJMJAkGQaSJAwDSRKGgSQJw0CShGEgScIwkCRhGEiSMAwkSRgGkiQM\nA0kShoEkCcNAkoRhIEnCMJAkYRhIkjAMJEkYBpIkDANJEoaBJIkWwiDJ6iT3JRlJcvFcb1+SdKA5\nDYMkC4BPAacDK4B1SVbMZQ+SpAPN9ZHBScBIVT1QVT8FrgPWzHEPkqT9zHUYLAEe7no92tQkSS1K\nVc3dxpLfA1ZX1Z80r88G3lBVF+4333pgffNyJbBzzpqc3xYB32u7iXnA/bCP+2If98U+x1fVy2ey\nwMLnqpODGAOWdb1e2tSepao2AhsBkgxX1dDctDe/uS863A/7uC/2cV/sk2R4psvM9WmibwHLkxyb\n5OeBtcCWOe5BkrSfOT0yqKonk1wIfAVYAGyqqrvnsgdJ0oHm+jQRVbUV2DqDRTY+V708D7kvOtwP\n+7gv9nFf7DPjfTGnF5AlSfOTX0chSZq/YeDXVnQkWZbk60nuSXJ3kova7qltSRYkuSPJf7TdS5uS\nvDLJDUm+neTeJL/Wdk9tSfLnzb+PnUk2J/mFtnuaK0k2JdmTZGdX7VVJtiX5TvN85FTrmZdh4NdW\nPMuTwAeqagVwMnDBYbwv9roIuLftJuaBTwBfrqpfAV7LYbpPkiwB/gwYqqqVdAanrG23qzl1FbB6\nv9rFwM1VtRy4uXl9SPMyDPBrK55RVeNVdXsz/SM6/+AP27u2kywF3gZ8pu1e2pTkF4E3A1cCVNVP\nq+qH7XbVqoXAS5IsBF4K/G/L/cyZqroFeHS/8hrg6mb6auDMqdYzX8PAr62YRJJB4ETgtnY7adU/\nAB8Enm67kZYdC0wA/9KcMvtMkiPabqoNVTUGfBT4LjAOPFZVX223q9YdVVXjzfQjwFFTLTBfw0D7\nSfIy4Ebg/VX1eNv9tCHJ24E9VbW97V7mgYXA64ANVXUi8GOmcSrghag5H76GTkAeDRyR5I/a7Wr+\nqM6Q0SmHjc7XMJjW11YcLpK8iE4QXFtVn2+7nxa9EXhHkl10Th3+VpJ/bbel1owCo1W19yjxBjrh\ncDh6C/BgVU1U1c+AzwO/3nJPbdudZDFA87xnqgXmaxj4tRWNJKFzXvjeqvpY2/20qaouqaqlVTVI\n57+J/6yqw/IvwKp6BHg4yfFN6VTgnhZbatN3gZOTvLT593Iqh+nF9C5bgHOa6XOAm6ZaYM7vQJ4O\nv7biWd4InA3clWRHU/tQcye3Dm9/Clzb/MH0AHBuy/20oqpuS3IDcDud0Xd3cBjdjZxkM3AKsCjJ\nKHApcDlwfZLzgIeAs6Zcj3cgS5Lm62kiSdIcMgwkSYaBJMkwkCRhGEiSMAwkSRgGkiQMA0kS8P+p\nltnH65e/+wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fa97bb6a668>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ax1 = plt.subplot()\n",
    "ax1.hist(distribution, bins=1000)\n",
    "_ = ax1.set_xlim([0, 10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Just to be sure, let's check how large are the \"windows\" we're going to look at our genes through"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "windows = [(gene, getGenecodeGenenameBoundary(gene)) for gene in genesToInvest]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "windowsSize = [(stop - start, start, stop, gene, chromosome, strand) for gene, (start, stop, chromosome, strand) in windows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "windowsSize.sort()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Then, let's check if there are any genes, which were incorrectly assigned very large windows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1697917, 78636716, 80334633, 'NRXN3', 'chr14', '+'),\n",
       " (1743269, 75955846, 77699115, 'ROBO2', 'chr3', '+'),\n",
       " (1783651, 67672276, 69455927, 'CTNNA3', 'chr10', '-'),\n",
       " (1825171, 55562531, 57387702, 'PCDH15', 'chr10', '-'),\n",
       " (1900278, 140988992, 142889270, 'LRP1B', 'chr2', '-'),\n",
       " (1987242, 64429876, 66417118, 'EYS', 'chr6', '-'),\n",
       " (2057827, 13976015, 16033842, 'MACROD2', 'chr20', '+'),\n",
       " (2059619, 2792875, 4852494, 'CSMD1', 'chr8', '-'),\n",
       " (2172911, 83166055, 85338966, 'DLG2', 'chr11', '-'),\n",
       " (2241764, 31115794, 33357558, 'DMD', 'chrX', '-'),\n",
       " (2298477, 8314246, 10612723, 'PTPRD', 'chr9', '-'),\n",
       " (2304996, 145813094, 148118090, 'CNTNAP2', 'chr7', '+'),\n",
       " (24961288, 29558488, 54519776, 'LINC01598', 'chr12', '-'),\n",
       " (95897243, 59208305, 155105548, 'DPH3P2', 'chrX', '+'),\n",
       " (95897327, 59311663, 155208990, 'ELOCP24', 'chrX', '-'),\n",
       " (95897873, 59318041, 155215914, 'TRPC6P', 'chrX', '-'),\n",
       " (95901878, 59353497, 155255375, 'WASH6P', 'chrX', '+'),\n",
       " (95910230, 59330252, 155240482, 'IL9R', 'chrX', '+'),\n",
       " (95911641, 59100480, 155012121, 'SPRY3', 'chrX', '+'),\n",
       " (95959471, 59213962, 155173433, 'VAMP7', 'chrX', '+')]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "windowsSize[-20:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Some reading suggests that everything larger than CNTNAP2 is due to errors in our annotation methodology or genecode and should not be included. (LINC01598 maybe isn't, but it's non-coding anyway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "maxWindowSizeThreshold = 5000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "minProbesetSize = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cleanedUpGenes = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for gene in genesToInvest:\n",
    "    start, stop, chromosome, strand = getGenecodeGenenameBoundary(gene)\n",
    "    if stop - start > maxWindowSizeThreshold:\n",
    "        continue\n",
    "    if len(getGenecodeGenenameToProbeset(gene)) < minProbesetSize:\n",
    "        continue\n",
    "    cleanedUpGenes.append(gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This leaves us with 23827 genes to look at"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23827"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cleanedUpGenes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# These functions are adapted from the earlier version of the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def pValueForProbeset(probeData, ma = None):\n",
    "    \"\"\"\n",
    "    computes the AS-statistic and the corresponing left and right p-value.\n",
    "    \"\"\"\n",
    "    left_pvalues = []\n",
    "    right_pvalues = []\n",
    "    if not ma:\n",
    "        ma = modalAllele()\n",
    "    for data in probeData:\n",
    "        regressR = linregress(ma, data)\n",
    "        left_pvalues.append(regressR.leftpvalue)\n",
    "        right_pvalues.append(regressR.rightpvalue)\n",
    "    leftresult = productCDF(product(left_pvalues), len(left_pvalues))\n",
    "    rightresult = productCDF(product(right_pvalues), len(right_pvalues))\n",
    "    return min(leftresult, rightresult) * 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def correctedData(gene, shuffleMA = False):\n",
    "    \"\"\"\n",
    "    computes the average gene expression correction.\n",
    "    \"\"\"\n",
    "    probesets = getGenecodeGenenameToProbeset(gene)\n",
    "    allPatientsGenes = [[] for _ in modalAllele()]\n",
    "    patientCorrections = []\n",
    "    for probeset in probesets:\n",
    "        for patientNo, probesetD in enumerate(probesetPatientData(probeset)):\n",
    "            for probeD in probesetD:\n",
    "                allPatientsGenes[patientNo].append(probeD)\n",
    "    for patientD in allPatientsGenes:\n",
    "        patientCorrections.append(sum(patientD)/len(patientD))\n",
    "    resultData = {}\n",
    "    pValues = {}\n",
    "    modalA = [i for i in modalAllele()]\n",
    "    if shuffleMA:\n",
    "        random.shuffle(modalA)\n",
    "    for probeset in probesets:\n",
    "        newData = []\n",
    "        for oldData in probesetData(probeset):\n",
    "            newData.append([point - correction for point, correction in zip(oldData, patientCorrections)])\n",
    "        resultData[probeset] = newData\n",
    "        pValue = pValueForProbeset(newData, ma = modalA)\n",
    "        pValues[probeset] = pValue\n",
    "    return (pValues, resultData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# An interesting observation, which shows that values greater that `0.9**4` are less likely than values less than `0.1**4` under the product distribution with 4 degrees of freedom.\n",
    "\n",
    "However, this won't influence our results, as the \"intermediate\" p-values we're using always come in pairs, `p` and `1-p`. After computing the product, we always take the smallest one, which makes the result consistent for both upregulated and downregulated exons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.000940523891325\n",
      "0.0182844955163\n"
     ]
    }
   ],
   "source": [
    "print(1 - productCDF(0.9**4, 4))\n",
    "print(productCDF(0.1**4, 4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Here we double-check that productCDF works (we expect answer to be close to 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.526882130193\n"
     ]
    }
   ],
   "source": [
    "k = 0\n",
    "v = 10\n",
    "for i in range(v):\n",
    "    k += productCDF(product([random.uniform(0, 1) for i in range(4)]), 4)\n",
    "print(k/v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import random\n",
    "geneData = {}\n",
    "for _, gene in zip(range(10), cleanedUpGenes):\n",
    "    geneData[gene] = correctedData(gene)[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Here we repeat the process of alternative splicing discovery from the previous pipeline, but this time using a null distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "nakamoriGenes = set()\n",
    "\n",
    "with open(\"geneList\") as f:\n",
    "    for line in f:\n",
    "        if not line:\n",
    "            continue\n",
    "        if line [0] == \"#\":\n",
    "            continue\n",
    "        gene = line.split(\",\")[0]\n",
    "        nakamoriGenes.add(gene)\n",
    "\n",
    "nakamoriNo = len(nakamoriGenes)\n",
    "commonGenes = nakamoriGenes.intersection(allOurGenes)\n",
    "agreedWith = len(commonGenes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We begin by counting how many probesets we have to look at."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "442050"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probesetCount = 0\n",
    "\n",
    "for gene in cleanedUpGenes:\n",
    "    probesetCount += len(getGenecodeGenenameToProbeset(gene))\n",
    "probesetCount"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We can then carry out the simulation to retrieve a \"real\" or \"final\" p-value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "redisConn[\"alterSplice\"].delete(\"simProbesets\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5710.0147793293\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "import time\n",
    "start = time.time()\n",
    "for i in range(10):\n",
    "    for _, gene in zip(range(2**64), cleanedUpGenes):\n",
    "        interPValues = correctedData(gene, shuffleMA = True)[0]\n",
    "        for (probeset, interPValue) in interPValues.items():\n",
    "            key = gene + \"$\" + probeset + \"$\" + str(i)\n",
    "            redisConn[\"alterSplice\"].zadd(\"simProbesets\", interPValue, key)\n",
    "stop = time.time()\n",
    "print(stop - start)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# As you can see, real p-values lose about 5 orders of magnitude."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.5649813369528334e-05\n",
      "0.0002397918787467481\n"
     ]
    }
   ],
   "source": [
    "def realPValue(interPValue):\n",
    "    lessThan = redisConn[\"alterSplice\"].zcount(\"simProbesets\", min=\"-inf\", max=interPValue)\n",
    "    total = redisConn[\"alterSplice\"].zcount(\"simProbesets\", min=\"-inf\", max=\"+inf\")\n",
    "    return lessThan/total\n",
    "\n",
    "print(realPValue(10**(-10)))\n",
    "print(realPValue(10**(-8)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# And finally we can run the experiment again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 399,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "redisConn[\"alterSplice\"].delete(\"simGeneAS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "609.6890335083008\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "for _, gene in zip(range(2**64), cleanedUpGenes):\n",
    "    interPValues = correctedData(gene)[0]\n",
    "    for (probeset, interPValue) in interPValues.items():\n",
    "        redisConn[\"alterSplice\"].zadd(\"simGeneAS\", realPValue(interPValue), gene + \"$\" + probeset)\n",
    "stop = time.time()\n",
    "print(stop - start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(b'ABLIM1$3307988', 4.524375070693361e-06),\n",
       " (b'ARHGEF9$4010527', 8.143875127248049e-06),\n",
       " (b'ABLIM1$3308051', 1.4478000226218754e-05),\n",
       " (b'ARHGAP5$3531485', 1.5382875240357426e-05),\n",
       " (b'RAPH1$2595566', 1.6966406515100102e-05),\n",
       " (b'ARHGAP5$3531527', 1.7192625268634768e-05),\n",
       " (b'ABLIM1$3308076', 1.7871281529238773e-05),\n",
       " (b'ABLIM1$3307940', 1.945481280398145e-05),\n",
       " (b'PKP2$3450255', 1.9907250311050785e-05),\n",
       " (b'TBX3$3472764', 1.9907250311050785e-05),\n",
       " (b'ABLIM1$3308075', 2.0133469064585454e-05),\n",
       " (b'ABLIM1$3308001', 2.171700033932813e-05),\n",
       " (b'ABLIM1$3308034', 2.375296912114014e-05),\n",
       " (b'ABLIM1$3308091', 2.420540662820948e-05),\n",
       " (b'ABLIM1$3307961', 2.488406288881348e-05),\n",
       " (b'ARHGAP5$3531521', 2.8729781698902838e-05),\n",
       " (b'NFIX$3822162', 2.9634656713041512e-05),\n",
       " (b'TBX3$3472777', 3.1444406741318854e-05),\n",
       " (b'PKP2$3450266', 3.2349281755457525e-05),\n",
       " (b'ABLIM1$3307963', 3.3027938016061534e-05),\n",
       " (b'ARHGAP5$3531482', 3.3027938016061534e-05)]"
      ]
     },
     "execution_count": 423,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "redisConn[\"alterSplice\"].zrange(\"simGeneAS\", 0, 20, withscores=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# In order to compare it with the result of the previous version of the pipeline we can take top 800 probesets and see how different the results are (just a sanity check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "allOurGenes = set()\n",
    "for geneProbesetScore in redisConn[\"alterSplice\"].zrange(\"simGeneAS\", 0, 800, withscores=True):\n",
    "    geneProbeset, score = geneProbesetScore\n",
    "    gene, probeset = d(geneProbeset).split(\"$\")\n",
    "    allOurGenes.add(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CHIC2', 'LIMCH1', 'CCDC14', 'F8', 'ZNF675', 'CDC42P3', 'TUBB1', 'SMAD2', 'STK24P1', 'LARP7', 'SGMS2', 'ZNF382', 'TMEM183A', 'ZNF667', 'CDNF', 'ZNF827', 'RAP1A', 'CCDC17', 'TBCD', 'GUSBP2', 'SEPT7P3', 'ZNF93', 'TBC1D23', 'LGR5', 'LRIG3', 'ZNF75A', 'CEP112', 'VPS35P1', 'ZNF721', 'FHOD1', 'RPGR', 'CILP', 'PAFAH1B2', 'SSFA2', 'MBD5', 'SAMD4A', 'MAP7D3', 'PPIC', 'CHMP5', 'ARHGAP5', 'ZNF778', 'CFHR1', 'PLPP1', 'C18orf32', 'NFIA', 'LIMS4', 'EOGT', 'EEF1A1P32', 'PHC3', 'NPIPB15', 'CHP1', 'ARHGEF9', 'ZDHHC2', 'TENM1', 'GUSBP1', 'SLC7A10', 'ABHD2', 'DMTF1', 'RPL23AP87', 'ATP11B', 'PDZD11', 'WASHC4', 'NF1', 'TBX3', 'ZNF626', 'NEDD4L', 'TBCK', 'HELZ', 'ABLIM1', 'DLG1', 'MTMR2', 'HMBOX1', 'NFIX', 'SOCS4', 'GLULP3', 'KCNQ1OT1', 'ZNF493', 'TTLL5', 'DIS3', 'SERF2', 'TM9SF2', 'STK3', 'PTPRD', 'RCBTB1', 'ZNF85', 'OFD1', 'PANK3', 'UBLCP1', 'EEF1A1P33', 'HMGB1', 'FGFRL1', 'FOXD4L1', 'SATB2', 'TPTE2', 'BPTF', 'TFPI', 'ZNF682', 'WDR75', 'RASA1', 'AATF', 'ZNF184', 'LNPEP', 'POLR1D', 'LINC00431', 'H3F3A', 'TUBB8P9', 'SMAD1', 'TRIM5', 'UGDH-AS1', 'PKP2', 'SNHG3', 'CSNK1G3', 'SYNE1', 'LIFR', 'TMEM256', 'CBWD1', 'C15orf57', 'GTF2IP1', 'DNMT1', 'TANC2', 'GINM1', 'ZNF385D', 'COPG2', 'ZNF254', 'MAST4', 'RAPH1', 'RBM25', 'TTLL1', 'RIMKLB', 'GUSBP4', 'ATP2A2', 'SLC24A3', 'CFHR3', 'ZNF92', 'CFH', 'FNDC3B', 'SMARCC1', 'COL3A1', 'GTF2I', 'FGD4', 'CKAP2', 'LRRC8B', 'PHLDA1', 'NDN', 'PAPOLA', 'MAPK1', 'G2E3', 'MBNL2', 'ZNF667-AS1', 'ZNF292', 'PAK2', 'OSTM1', 'ZNF429', 'DNAJB11', 'PROSER1', 'ZNF355P', 'SIPA1', 'EMC1', 'OSBPL8', 'AEBP2', 'FBXO31', 'UPF3A', 'ERGIC2', 'CCDC93', 'NIPBL', 'LPP', 'GGNBP2', 'SESN3', 'ZNF732', 'RCAN3', 'GOLT1B', 'ZEB1', 'ZNHIT6', 'ANKRD10', 'MAP3K7', 'NCK2', 'ZNF257', 'PGRMC2', 'CLCN1', 'CSNK2A2', 'LIMS1', 'UBE2Q2', 'NCAM1', 'NHLRC2', 'ABLIM2', 'SRSF8', 'PICALM', 'UBE2J1', 'NPIPB3', 'FRG1BP', 'CFHR4', 'ARHGEF12', 'FAM227B', 'PALMD', 'SAE1', 'ZCCHC6', 'ZNF90P1', 'ZNF676', 'ZNF518A', 'TCP11L2', 'TPTE2P4', 'ZC3H6', 'UBXN7', 'PNN', 'UMAD1', 'OR7E136P', 'EIF3A', 'MYLK2', 'NCSTN', 'CDC42BPB', 'GBP2', 'NAIP', 'TMX3', 'TBC1D15', 'NPIPA5', 'CCDC88A', 'NFIB', 'ROCK1', 'ZNF430', 'ANKRD12', 'TERF1P5', 'CHD9', 'LMAN1', 'TAF2', 'SRP72P1', 'FAM49B', 'CDC73', 'ENPEP', 'MBIP', 'ERRFI1', 'PEG10', 'SLC25A13', 'R3HDM2', 'OBSCN', 'ZNF727', 'STRN', 'CAST', 'ATG3', 'PHF21A', 'DAPK2', 'C3', 'SUPT3H', 'ARFGEF1', 'PLRG1', 'LRR1', 'CEP57', 'SEPT7', 'UBA6', 'ADRB2', 'C5orf24', 'PPP1R12A', 'ADAM10', 'EEA1', 'C9orf85', 'PRKD1', 'VPS35', 'C2CD3', 'ZNF468', 'ZNF880', 'PHF7', 'CBWD3', 'PIGK', 'TNPO1P1', 'NPIPB14P', 'KDM4C', 'TET1', 'TMEM9B', 'NPIPA7', 'IMPAD1', 'TMX1', 'DOCK9', 'PREPL', 'HAUS2', 'NPIPB11', 'C1orf27', 'UTRN', 'LUM', 'GRB14', 'LDB3', 'LRRC75A-AS1', 'YAP1P1', 'DAPK1', 'FAM120B', 'ZSWIM7', 'CLDND1', 'SEC62', 'PTOV1-AS1', 'C9orf24', 'OR7E46P', 'RSU1', 'NKTR', 'CFAP61', 'AKAP13', 'FAT4', 'EXOSC10', 'KIF2A', 'BBS10', 'NFAT5', 'RNF2P1', 'KATNBL1', 'TSTD1', 'RXFP2', 'ANKRD36', 'TET2', 'NEMP1', 'NEK10', 'ANGPT1', 'ILF3', 'CAPRIN2', 'PLEKHA1', 'PSMA2', 'MXRA7', 'LINC01697', 'ZFC3H1', 'PARP11', 'ANP32A', 'EEF1A1P12', 'TMEM47', 'SLC25A17', 'ATP1A1', 'GTF2IP4'}\n"
     ]
    }
   ],
   "source": [
    "print(allOurGenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "nakamoriGenes = set()\n",
    "with open(\"geneList\") as f:\n",
    "    for line in f:\n",
    "        if not line:\n",
    "            continue\n",
    "        if line [0] == \"#\":\n",
    "            continue\n",
    "        gene = line.split(\",\")[0]\n",
    "        nakamoriGenes.add(gene)\n",
    "\n",
    "nakamoriNo = len(nakamoriGenes)\n",
    "commonGenes = nakamoriGenes.intersection(allOurGenes)\n",
    "agreedWith = len(commonGenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 412,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ATP2A2', 'CLCN1', 'FHOD1', 'LDB3', 'MBNL2', 'NFIX', 'TBC1D15'}"
      ]
     },
     "execution_count": 412,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "commonGenes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We still agree on 7 genes, but this time ATP2A2 makes it to the list, while MBNL1 doesn't."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# We can also look at the zinc fingers hypothesis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 406,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x, M, n, N] = [agreedWith - 1, len(cleanedUpGenes), nakamoriNo , len(allOurGenes)]\n",
    "scipy.stats.hypergeom.sf(x, M, n, N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "geneExpression = set()\n",
    "with open(\"geneExpression\") as ge:\n",
    "    for line in ge:\n",
    "        for elem in line.rstrip().split():\n",
    "            geneExpression.add(elem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "321\n",
      "877\n",
      "25\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.00037788473593942858"
      ]
     },
     "execution_count": 408,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "commonGE = allOurGenes.intersection(geneExpression)\n",
    "print(len(allOurGenes))\n",
    "print(len(geneExpression))\n",
    "print(len(commonGE))\n",
    "[x, M, n, N] = [len(commonGE) - 1, len(cleanedUpGenes), len(geneExpression) , len(allOurGenes)]\n",
    "scipy.stats.hypergeom.sf(x, M, n, N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "transvis",
   "language": "python",
   "name": "transvis"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
